### PR TITLE
Sub: updates to Position Hold mode

### PR DESCRIPTION
**Pull request by Williangalvani**
This switches Poshold to use the PosControl library and same depth hold logic as used in the depth-hold mode.

### CODE REVIEW

File: ArduSub/control_poshold.cpp [hunk 1]
```diff
@@ -111,19 +115,36 @@ void Sub::poshold_run()
 
     ///////////////////
     // Update z axis //
+    // Same logic as in control_althold.cpp
```
Comment (jaxxzer) [inline ArduSub/control_poshold.cpp:118]: We should pull this out into a common function called by both modes

File: ArduSub/control_althold.cpp [hunk 2]
```diff
@@ -96,7 +96,14 @@ void Sub::althold_run()
...
 
-    // Hold actual position until zero derivative is detected
+    stabilize_depth();
+
+    motors.set_forward(channel_forward->norm_input());
+    motors.set_lateral(channel_lateral->norm_input());
+}
+
+void Sub::stabilize_depth() {
+        // Hold actual position until zero derivative is detected
```
Comment (patrickelectric) [inline ArduSub/control_althold.cpp:106]: wrong indentation

File: ArduSub/control_poshold.cpp [hunk 3]
```diff
@@ -22,10 +25,9 @@ bool Sub::poshold_init()
     pos_control.set_alt_target(inertial_nav.get_altitude());
     pos_control.set_desired_velocity_z(inertial_nav.get_velocity_z());
 
-    // set target to current position
-    // only init here as we can switch to PosHold in flight with a velocity <> 0 that will be used as _last_vel in PosControl and never updated again as we inhibit Reset_I
-    loiter_nav.clear_pilot_desired_acceleration();
-    loiter_nav.init_target();
+    attitude_control.set_throttle_out(0.5 ,true, g.throttle_filt);
```
Comment (jaxxzer) [inline ArduSub/control_poshold.cpp:28]: 0.5f
please briefly comment these hunks

File: ArduSub/control_poshold.cpp [hunk 4]
```diff
@@ -63,19 +62,24 @@ void Sub::poshold_run()
...
+    pos_control.set_desired_velocity_xy(0,0);
+
+    // TODO: check if 'visual_odom.healthy' is required for real hardware
+    if (visual_odom.healthy() || position_ok()) {
+        // Allow pilot to reposition the sub
+        if (fabsf(pilot_lateral) > 0.1 || fabsf(pilot_forward) > 0.1) {
+            pos_control.set_target_to_stopping_point_xy();
+        }
+        translate_pos_control_rp(lateral_out, forward_out);
+        // run loiter controller
```
Comment (jaxxzer) [inline ArduSub/control_poshold.cpp:74]: stale

File: Tools/autotest/ardusub.py [hunk 5]
```diff
@@ -147,6 +147,52 @@ def test_alt_hold(self):
...
+        self.arm_vehicle()
+        # point North
+        self.reach_heading_manual(0)
+        self.mavproxy.send('mode POSHOLD\n')
+        self.wait_mode('POSHOLD')
+
+        #dive a little 
+        self.set_rc(Joystick.Throttle, 1300)
+        self.delay_sim_time(3)
+        self.set_rc(Joystick.Throttle, 1500)
```
Comment (jaxxzer) [inline Tools/autotest/ardusub.py:162]: should we wait for deceleration?

File: Tools/autotest/default_params/sub.parm [hunk 6]
```diff
@@ -58,9 +58,10 @@ INS_ACC3OFFS_Z   0.000
 INS_ACC3SCAL_X   1.000
 INS_ACC3SCAL_Y   1.000
 INS_ACC3SCAL_Z   1.000
-PSC_ACCZ_P   4
-PSC_VECZ_P   8
-PSC_POSZ_P   3
+PSC_POSXY_P 2.5
```
Comment (jaxxzer) [inline Tools/autotest/default_params/sub.parm:61]: why are we replacing z with xy, instead of appending?

File: ArduSub/control_althold.cpp [hunk 7]
```diff
@@ -125,7 +132,4 @@ void Sub::althold_run()
 
         pos_control.update_z_controller();
     }
-
-    motors.set_forward(channel_forward->norm_input());
-    motors.set_lateral(channel_lateral->norm_input());
-}
+}
```
Comment (jaxxzer) [inline ArduSub/control_althold.cpp:135]: whitespace

File: ArduSub/control_poshold.cpp [hunk 8]
```diff
@@ -63,19 +62,24 @@ void Sub::poshold_run()
...
-    // Allow pilot to reposition the sub
-    if (fabsf(pilot_lateral) > 0.1 || fabsf(pilot_forward) > 0.1) {
-        lateral_out = pilot_lateral;
-        forward_out = pilot_forward;
-        loiter_nav.clear_pilot_desired_acceleration();
-        loiter_nav.init_target(); // initialize target to current position after repositioning
+    pos_control.set_desired_velocity_xy(0,0);
+
+    // TODO: check if 'visual_odom.healthy' is required for real hardware
+    if (visual_odom.healthy() || position_ok()) {
```
Comment (jaxxzer) [inline ArduSub/control_poshold.cpp:68]: i think this should be only position_ok(). if necessary, we can add the visual_odom healthy check to Sub::optflow_position_ok logic

File: Tools/autotest/default_params/sub.parm (hunk 6 above)
Comment (Williangalvani) [reply Tools/autotest/default_params/sub.parm:61]: oops

File: ArduSub/control_poshold.cpp (hunk 8 above)
Comment (Williangalvani) [reply ArduSub/control_poshold.cpp:68]: It was just my memory failing on me, it works with position_ok() now:
https://github.com/ArduPilot/ardupilot/pull/13982#issuecomment-609921878

File: ArduSub/control_poshold.cpp [hunk 9]
```diff
@@ -63,19 +63,22 @@ void Sub::poshold_run()
...
-    if (fabsf(pilot_lateral) > 0.1 || fabsf(pilot_forward) > 0.1) {
-        lateral_out = pilot_lateral;
-        forward_out = pilot_forward;
-        loiter_nav.clear_pilot_desired_acceleration();
-        loiter_nav.init_target(); // initialize target to current position after repositioning
+    pos_control.set_desired_velocity_xy(0,0);
+
+    if (position_ok()) {
+        // Allow pilot to reposition the sub
+        if (fabsf(pilot_lateral) > 0.1 || fabsf(pilot_forward) > 0.1) {
```
Comment (jaxxzer) [inline ArduSub/control_poshold.cpp:70]:
```suggestion
        if (fabsf(pilot_lateral) > 0.1f || fabsf(pilot_forward) > 0.1f) {
```

**Pull request merged**